features/api-layer/src/main/java/org/opennms/features/apilayer/pollers/PollerRequestBuilderImpl.java

```diff
@@ -0,0 +1,178 @@
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.apilayer.pollers;
+
+import java.net.InetAddress;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import org.opennms.integration.api.v1.pollers.PollerRequestBuilder;
+import org.opennms.integration.api.v1.pollers.PollerResult;
+import org.opennms.integration.api.v1.pollers.Status;
+import org.opennms.netmgt.poller.LocationAwarePollerClient;
+import org.opennms.netmgt.poller.MonitoredService;
+import org.opennms.netmgt.poller.PollStatus;
+import org.opennms.netmgt.poller.PollerResponse;
+import org.opennms.netmgt.poller.support.SimpleMonitoredService;
+
+/**
+ * Builder Implementation for {@link org.opennms.integration.api.v1.pollers.ServicePollerClient}
+ */
+public class PollerRequestBuilderImpl implements PollerRequestBuilder {
+
+    public PollerRequestBuilderImpl(LocationAwarePollerClient pollerClient) {
+        this.pollerClient = pollerClient;
+    }
+
+    private String className;
+
+    private InetAddress address;
+
+    private String serviceName;
+
+    private Long ttlInMs;
+
+    private Map<String, String> attributes = new HashMap<>();
+
+    private final LocationAwarePollerClient pollerClient;
+
+
+    @Override
+    public PollerRequestBuilder withPollerClassName(String className) {
+        this.className = className;
+        return this;
+    }
+
+    @Override
+    public PollerRequestBuilder withAddress(InetAddress address) {
+        this.address = address;
+        return this;
+    }
+
+    @Override
+    public PollerRequestBuilder withServiceName(String serviceName) {
+        this.serviceName = serviceName;
+        return this;
+    }
+
+    @Override
+    public PollerRequestBuilder withAttribute(String key, String value) {
+        this.attributes.put(key, value);
+        return this;
+    }
+
+    @Override
+    public PollerRequestBuilder withAttributes(Map<String, String> attributes) {
+        this.attributes.putAll(attributes);
+        return this;
+    }
+
+    @Override
+    public PollerRequestBuilder withTimeToLive(Long ttlInMs) {
+        this.ttlInMs = ttlInMs;
+        return this;
+    }
+
+    @Override
+    public CompletableFuture<PollerResult> execute() {
```

Comment:
Can the `execute()` method be moved out of the request and into the client? I think that would be better encapsulation and we could avoid having to pass in an instance of the client to the request.

The request builder could then become an actual builder that returns a request object which is accepted as a parameter in the `execute()` method that we have moved (or otherwise if you have a better idea). This will involve updating the api interfaces.

Reply:
yeah, here we are using builder for building and executing as well.  That allows us to do 
`poll().withService().withClassName()...execute()`.    By separating builder we need to build first and then executes the request.  Also it needs to know actual instance of the builder not just the interface.   I guess the current one makes it easier for the clients to call without actually bothering about the implementation.   This follows the pattern that we used for all `LocationAware` builders.

Reply:
You could still maintain the same call signature if you changed the `execute()` inside the builder to just be a proxy for the client that built it's `execute()`.

So the builder would then take a reference to the `ServicePollerClient` that created it on construction rather than a reference to a `LocationAwarePollerClient`. Inside its execute it would simply pass itself to the execute method of the `ServicePollerClient`. This would move all the execution logic inside the client class which seems like better encapsulation to me.

Perhaps an even better solution would be to move the builder class into the client class as a nested class and then you could avoid passing the reference to the client altogether.

Feel free to disagree as I'm not very familiar with this code.

Example change to `ServicePollerClientImpl`:
```
   @Override
    public PollerRequestBuilder poll() {
        return new PollerRequestBuilderImpl(this);
    }

    CompletableFuture<PollerResult> execute(PollerRequestBuilder builder) {
        ... existing execute logic
    }
```

Example change to `PollerRequestBuilderImpl `: 
```
    public PollerRequestBuilderImpl(ServicePollerClient pollerClient) {
        this.pollerClient = pollerClient;
    }

    @Override
    public CompletableFuture<PollerResult> execute() {
        return pollerClient.execute(this);
    }
```

Reply:
@cgorantla Not sure if you are still looking at this, but feel free to ignore this suggestion as I went off on a bit of a tangent here, the original impl is fine especially if it is attempting to remain consistent with the existing example.